openhcl/virt_mshv_vtl/src/processor/hardware_cvm/mod.rs

```diff
@@ -1195,6 +1216,65 @@ impl<B: HardwareIsolatedBacking> UhProcessor<'_, B> {
 
         Ok(reprocessing_required)
     }
+
+    fn get_vsm_vp_secure_config_vtl(
+        &mut self,
+        requesting_vtl: GuestVtl,
+        target_vtl: GuestVtl,
+    ) -> Result<HvRegisterVsmVpSecureVtlConfig, HvError> {
+        if requesting_vtl <= target_vtl {
+            return Err(HvError::AccessDenied);
+        }
+
+        let requesting_vtl = requesting_vtl.into();
+
+        let guest_vsm_lock = self.partition.guest_vsm.read();
+        let guest_vsm = guest_vsm_lock
+            .get_hardware_cvm()
+            .ok_or(HvError::InvalidVtlState)?;
+
+        let tlb_locked = self.vtls_tlb_locked.get(requesting_vtl, target_vtl);
+
+        Ok(HvRegisterVsmVpSecureVtlConfig::new()
+            .with_mbec_enabled(guest_vsm.mbec_enabled)
+            .with_tlb_locked(tlb_locked))
+    }
+
+    fn set_vsm_vp_secure_config_vtl(
+        &mut self,
+        requesting_vtl: GuestVtl,
+        target_vtl: GuestVtl,
+        config: HvRegisterVsmVpSecureVtlConfig,
+    ) -> Result<(), HvError> {
+        if requesting_vtl <= target_vtl {
+            return Err(HvError::AccessDenied);
+        }
+
+        if config.supervisor_shadow_stack_enabled() || config.hardware_hvpt_enabled() {
+            return Err(HvError::InvalidRegisterValue);
+        }
+
+        let requesting_vtl = requesting_vtl.into();
+
+        let guest_vsm_lock = self.partition.guest_vsm.read();
+        let guest_vsm = guest_vsm_lock
+            .get_hardware_cvm()
+            .ok_or(HvError::InvalidVtlState)?;
+
+        // MBEC must always be enabled or disabled partition-wide.
+        if config.mbec_enabled() != guest_vsm.mbec_enabled {
+            return Err(HvError::InvalidRegisterValue);
+        }
+
+        let tlb_locked = self.vtls_tlb_locked.get(requesting_vtl, target_vtl);
+        match (tlb_locked, config.tlb_locked()) {
+            (true, false) => self.unlock_tlb_lock(requesting_vtl),
```

Comment:
I don't think this is quite right, since this would unlock for all lower vtls, not just the target vtl...let me think...

Reply:
I guess this is fine for now, given we only support 1 and 0, but this would be doing too much in some hypothetical future. We'd have to pull out the for loop inside unlock_tlb_lock so that this can unlock just one target. Up to you if you wanna do that now or not.